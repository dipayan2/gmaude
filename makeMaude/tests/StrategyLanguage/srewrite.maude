set show timing off .
set show advisories off .

mod DATA is
  protecting SET{Qid} .

  sort Pair .

  op <_|_> : Set{Qid} Set{Qid} -> Pair [ctor] .

  vars Q Q1 Q2 : Qid .

  rl [concat] : Q1, Q2 => qid(string(Q1) + string(Q2)) .
  rl [rh] : Q => qid("h" + string(Q)) .

  rl [rm] : Q => empty .
endm


smod STRATEGIES is
  protecting DATA .

  var Q      : Qid .
  var S S1 S2   : Set{Qid} .

  strat filter : Set{Qid} @ Set{Qid} .

  sd filter(S1) := match empty .
  sd filter(S1) := try(matchrew Q, S s.t. Q in S1 by Q using rm[Q <- Q], S using filter(S1)) .
endsm

srew < 'a, 'b | 'c, 'd, 'e > using matchrew < S1 | S2 > by S1 using rh , S2 using concat ! .
srew < 'a, 'b | 'a, 'd > using matchrew < S1 | S2 > by S1 using one(concat) , S2 using filter(S1) .

srew [2] < 'a | 'c > using matchrew < S1 | S2 > by S1 using rh * , S2 using rh * .
continue 1 .

***
*** Testing all the combinators
***

fmod 15PUZZLE-BOARD is
  protecting NAT .

  sorts Tile Row Puzzle .
  subsorts Nat < Tile < Row < Puzzle .

  op b : -> Tile [ctor] .
  op nil : -> Row [ctor] .
  op __  : Row Row -> Row [ctor assoc id: nil prec 25] .
  op _;_ : Puzzle Puzzle -> Puzzle [ctor assoc] .

  var T : Tile .
  var R : Row .

  op size : Row -> Nat .
  eq size(nil) = 0 .
  eq size(T R) = size(R) + 1 .
endfm

mod 15PUZZLE is
  protecting 15PUZZLE-BOARD .
  var T : Tile .
  vars LU RU LD RD : Row .
  var P : Puzzle .
   rl [left]  : T b => b T .
   rl [right] : b T => T b .
  crl [down]  : (LU b RU) ; (LD T RD)
             => (LU T RU) ; (LD b RD) if size(LU) = size(LD) .
  crl [up]    : (LU T RU) ; (LD b RD)
             => (LU b RU) ; (LD T RD) if size(LU) = size(LD) .
endm

srewrite 1 b 2 using idle .
srewrite 1 b 2 using fail .
srewrite 1 b 2 ; 3 b 4 using right .
srewrite 1 b 2 ; 3 b 4 using left[T <- 1] .

mod 15PUZZLE-LOG is
  protecting 15PUZZLE .
  protecting LIST{Qid} .
  sort PuzzleLog .
  op <_|_> : List{Qid} Puzzle -> PuzzleLog [ctor] .
  var M : Qid .
  var L : List{Qid} .
  vars P P' : Puzzle .
  crl [move] : < L | P > => < L M | P' > if P => P' [nonexec] .
endm

srewrite < nil | 1 b 2 > using move[M <- 'left]{left} .
srewrite < nil | 1 b 2 > using all .
srewrite 1 b 2 using xmatch b N:Nat s.t. N:Nat =/= 1 .
srewrite 1 2 ; 3 b using left ; up | up ; left .
srewrite 1 b 2 3 using right * .
srewrite 1 b 2 3 4 using right * ; (right ? fail : idle) .
srewrite 1 b 2 ; 3 b 4 using matchrew RU:Row ; RD:Row by RU:Row using left, RD:Row using right .
srewrite 1 b 2 3 using one(right +) .
srew [2] b 1 2 ; b 3 using right + .
dsrew [2] b 1 2 ; b 3 using right + .
srew 1 b using (left | right) * .

smod 15PUZZLE-STRATS is
  protecting 15PUZZLE .
  protecting INT .
  strat loop           @ Puzzle .
  strat move : Int Int @ Puzzle .
  var N : Nat .
  var M : Int .
  sd loop := left ; up ; right ; down .
  sd move(0, 0)      := idle .
  sd move(s(N), M)   := right ; move(N, M) .
  sd move(- s(N), M) := left  ; move(- N, M) .
  sd move(0, s(N))   := down  ; move(0, N) .
  sd move(0, - s(N)) := up    ; move(0, - N) .
endsm

srewrite 1 2 3 ; 4 5 6 ; 7 b 8 using move(1, -2) .

***
*** Testing more complex strategy definitions
***

mod SWAPPING{X :: DEFAULT} is
  protecting ARRAY{Nat, X} .
  vars I J : Nat .
  vars V W : X$Elt .
  var AR : Array{Nat, X} .
  rl [swap] : J |-> V ; I |-> W => J |-> W ; I |-> V .

  op maxIndex : Array{Nat, X} ~> Nat .
  eq maxIndex(empty) = 0 .
  eq maxIndex(I |-> V ; AR) = if maxIndex(AR) < I then I else maxIndex(AR) fi .
endm

view DEFAULT+ from DEFAULT to STRICT-TOTAL-ORDER + DEFAULT is
endv

smod INSERTION-SORT{X :: STRICT-TOTAL-ORDER + DEFAULT} is
  protecting SWAPPING{DEFAULT+}{X} * (
    sort Array{Nat, DEFAULT+}{X} to NatArray{X}
  ) .

  strat  swap : Nat Nat @ NatArray{X} .
  strats insert insort : Nat @ NatArray{X} .

  vars X Y J I : Nat .
  vars V W : X$Elt .
  var AR : NatArray{X} .

  sd insort(Y) := try(match AR s.t. Y <= maxIndex(AR) ;
                      insert(Y) ;
                      insort(Y + 1)) .

  sd insert(1) := idle [label base-case] .
 csd insert(s(X)) := try(xmatch X |-> V ; s(X) |-> W s.t. W < V ;
                         swap(X, s(X)) ;
                         insert(X))
                      if X > 0 [label recursive-case] .

  sd swap(X, Y) := swap[J <- X, I <- Y] .
endsm

view Int<0 from STRICT-TOTAL-ORDER + DEFAULT to INT is
  sort Elt to Int .
endv

smod INSERTION-SORT-INT is
  protecting INSERTION-SORT{Int<0} .
endsm

srewrite 1 |-> 8 ; 2 |-> 3 ; 3 |-> 15 ; 4 |-> 5 ; 5 |-> 2 using insort(2) .
