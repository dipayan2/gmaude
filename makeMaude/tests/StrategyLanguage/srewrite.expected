==========================================
srewrite in STRATEGIES : < 'a, 'b | 'c, 'd, 'e > using matchrew < S1 | S2 > by
    S1 using rh, S2 using concat ! .

Solution 1
rewrites: 93
result Pair: < 'b, 'ha | 'cde >

Solution 2
rewrites: 98
result Pair: < 'a, 'hb | 'cde >

Solution 3
rewrites: 98
result Pair: < 'b, 'ha | 'ecd >

Solution 4
rewrites: 98
result Pair: < 'b, 'ha | 'ced >

Solution 5
rewrites: 98
result Pair: < 'a, 'hb | 'ecd >

Solution 6
rewrites: 98
result Pair: < 'a, 'hb | 'ced >

Solution 7
rewrites: 98
result Pair: < 'b, 'ha | 'dce >

Solution 8
rewrites: 98
result Pair: < 'a, 'hb | 'dce >

Solution 9
rewrites: 98
result Pair: < 'b, 'ha | 'edc >

Solution 10
rewrites: 98
result Pair: < 'a, 'hb | 'edc >

Solution 11
rewrites: 98
result Pair: < 'b, 'ha | 'dec >

Solution 12
rewrites: 98
result Pair: < 'a, 'hb | 'dec >

No more solutions.
rewrites: 98
==========================================
srewrite in STRATEGIES : < 'a, 'b | 'a, 'd > using matchrew < S1 | S2 > by S1
    using one(concat), S2 using filter(S1) .

Solution 1
rewrites: 9
result Pair: < 'ab | 'd >

No more solutions.
rewrites: 9
==========================================
srewrite [2] in STRATEGIES : < 'a | 'c > using matchrew < S1 | S2 > by S1 using
    rh *, S2 using rh * .

Solution 1
rewrites: 8
result Pair: < 'a | 'c >

Solution 2
rewrites: 16
result Pair: < 'ha | 'c >

Solution 3
rewrites: 0
result Pair: < 'a | 'hc >
==========================================
srewrite in 15PUZZLE : 1 b 2 using idle .

Solution 1
rewrites: 0
result Row: 1 b 2

No more solutions.
rewrites: 0
==========================================
srewrite in 15PUZZLE : 1 b 2 using fail .

No solution.
rewrites: 0
==========================================
srewrite in 15PUZZLE : 1 b 2 ; 3 b 4 using right .

Solution 1
rewrites: 1
result Puzzle: 1 2 b ; 3 b 4

Solution 2
rewrites: 2
result Puzzle: 1 b 2 ; 3 4 b

No more solutions.
rewrites: 2
==========================================
srewrite in 15PUZZLE : 1 b 2 ; 3 b 4 using left[T <- 1] .

Solution 1
rewrites: 1
result Puzzle: b 1 2 ; 3 b 4

No more solutions.
rewrites: 1
==========================================
srewrite in 15PUZZLE-LOG : < nil | 1 b 2 > using move[M <- 'left]{left} .

Solution 1
rewrites: 2
result PuzzleLog: < 'left | b 1 2 >

No more solutions.
rewrites: 2
==========================================
srewrite in 15PUZZLE-LOG : < nil | 1 b 2 > using all .

Solution 1
rewrites: 1
result PuzzleLog: < nil | b 1 2 >

Solution 2
rewrites: 2
result PuzzleLog: < nil | 1 2 b >

No more solutions.
rewrites: 2
==========================================
srewrite in 15PUZZLE-LOG : 1 b 2 using xmatch b N:Nat such that N:Nat =/= 1 =
    true .

Solution 1
rewrites: 1
result Row: 1 b 2

No more solutions.
rewrites: 1
==========================================
srewrite in 15PUZZLE-LOG : 1 2 ; 3 b using left ; up | up ; left .

Solution 1
rewrites: 15
result Puzzle: b 2 ; 1 3

Solution 2
rewrites: 20
result Puzzle: b 1 ; 3 2

No more solutions.
rewrites: 20
==========================================
srewrite in 15PUZZLE-LOG : 1 b 2 3 using right * .

Solution 1
rewrites: 0
result Row: 1 b 2 3

Solution 2
rewrites: 1
result Row: 1 2 b 3

Solution 3
rewrites: 2
result Row: 1 2 3 b

No more solutions.
rewrites: 2
==========================================
srewrite in 15PUZZLE-LOG : 1 b 2 3 4 using right * ; (right ? fail : idle) .

Solution 1
rewrites: 6
result Row: 1 2 3 4 b

No more solutions.
rewrites: 6
==========================================
srewrite in 15PUZZLE-LOG : 1 b 2 ; 3 b 4 using matchrew RU:Row ; RD:Row by
    RU:Row using left, RD:Row using right .

Solution 1
rewrites: 2
result Puzzle: b 1 2 ; 3 4 b

No more solutions.
rewrites: 2
==========================================
srewrite in 15PUZZLE-LOG : 1 b 2 3 using one(right +) .

Solution 1
rewrites: 1
result Row: 1 2 b 3

No more solutions.
rewrites: 1
==========================================
srewrite [2] in 15PUZZLE-LOG : b 1 2 ; b 3 using right + .

Solution 1
rewrites: 2
result Puzzle: 1 b 2 ; b 3

Solution 2
rewrites: 3
result Puzzle: b 1 2 ; 3 b
==========================================
dsrewrite [2] in 15PUZZLE-LOG : b 1 2 ; b 3 using right + .

Solution 1
rewrites: 1
result Puzzle: 1 b 2 ; b 3

Solution 2
rewrites: 2
result Puzzle: 1 2 b ; b 3
==========================================
srewrite in 15PUZZLE-LOG : 1 b using (left | right)* .

Solution 1
rewrites: 0
result Row: 1 b

Solution 2
rewrites: 1
result Row: b 1

No more solutions.
rewrites: 2
==========================================
srewrite in 15PUZZLE-STRATS : 1 2 3 ; 4 5 6 ; 7 b 8 using move(1, -2) .

Solution 1
rewrites: 52
result Puzzle: 1 2 b ; 4 5 3 ; 7 8 6

No more solutions.
rewrites: 52
==========================================
srewrite in INSERTION-SORT-INT : 1 |-> 8 ; 2 |-> 3 ; 3 |-> 15 ; 4 |-> 5 ; 5 |->
    2 using insort(2) .

Solution 1
rewrites: 116
result NatArray{Int<0}: 1 |-> 2 ; 2 |-> 3 ; 3 |-> 5 ; 4 |-> 8 ; 5 |-> 15

No more solutions.
rewrites: 116
Bye.
